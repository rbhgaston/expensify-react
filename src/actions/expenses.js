// import {v4 as uuidv4} from 'uuid';
import db from '../firebase/firebase'

// ADD
export const  addExpense = (expense) => ({
    type: 'ADD',
    expense
})

export const startAddExpense = ({description= '', note= '', createdAt= 0, amount= 0} = {}) =>
    (dispatch, getState) => {
        const uid = getState().auth.uid
        const expense = {
            description,
            amount,
            note,
            createdAt
        }
        return db.ref(`users/${uid}/expenses`).push(expense)
            .then(ref => dispatch(addExpense({
                id: ref.key,
                ...expense
            })))
    }

// EDIT
export const editExpense = (id, updates) => ({
    type: 'EDIT',
    id,
    updates
})

export const startEditExpense= (id, updates) =>
    (dispatch, getState) => {
    const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses/${id}`).update(updates)
            .then(() => dispatch(editExpense(id, updates)))
    }



// REMOVE
export const removeExpense = (id) => ({type: 'REMOVE', id})

export const startRemoveExpense = (id) =>
    (dispatch, getState) => {
    const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses/${id}`).remove()
            .then(() => dispatch(removeExpense(id)))
        }

// SET
export const setExpenses = (expenses) => ({
    type: 'SET',
    expenses
})

export const startSetExpenses = () =>
    (dispatch, getState) => {
        const uid = getState().auth.uid
        return db.ref(`users/${uid}/expenses`).once('value').then(snapshot => {
            const expensesArray = []
            snapshot.forEach(expense => {expensesArray.push({id: expense.key.toString(), ...expense.val()})})
            dispatch(setExpenses(expensesArray))
        })
    }