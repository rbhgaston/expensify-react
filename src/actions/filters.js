export const sortByAmount = () =>  ({type:'AMOUNT'})
export const sortByDate = () => ({type: 'DATE'})
export const setTextFilter = (text = '') => ({type: 'TEXT', text})

// dates are moment objects
export const setStartDate = (startDate) => ({type: 'START_DATE', startDate})
export const setEndDate = (endDate) => ({type: 'END_DATE', endDate})