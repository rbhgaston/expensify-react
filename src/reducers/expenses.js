const expensesReducerDefaultState = []
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD':
            return [
                ...state,
                action.expense
            ]
        case 'EDIT':
            return state.map(expense => expense.id === action.id ? {...expense, ...action.updates} : expense )
        case 'REMOVE':
            return state.filter(expense => expense.id !== action.id)
        default:
            return state
    }
}

export default expensesReducer