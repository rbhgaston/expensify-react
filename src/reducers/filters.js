import moment from 'moment';

const filtersReducerDefaultState = {
    text: '',
    sort: 'amount',
    startDate: moment().startOf('month'),
    endDate: moment().endOf('month')
}
const filtersReducer = (state = filtersReducerDefaultState, action) => {
    switch (action.type) {
        case 'AMOUNT':
            return {
                ...state,
                sort: 'amount'
            }
        case 'DATE':
            return{
                ...state,
                sort: 'date'
            }
        case 'TEXT':
            return {
                ...state,
                text: action.text
            }
        case 'START_DATE':
            return {
                ...state,
                startDate: action.startDate
            }
        case 'END_DATE':
            return {
                ...state,
                endDate: action.endDate
            }
        default:
            return state
    }
}

export default filtersReducer