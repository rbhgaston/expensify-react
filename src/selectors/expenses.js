import moment from 'moment';

export const getFilteredExpenses =  (expenses, {text, sort, startDate, endDate}) => {
    return expenses.filter(expense => {
        const isGreaterThanStart = startDate ?  moment(expense.createdAt).isSameOrAfter(startDate, 'day') : true
        const isLessThanEnd = endDate ? moment(expense.createdAt).isSameOrBefore(endDate, 'day') : true

        return expense.description.includes(text) && isGreaterThanStart && isLessThanEnd
        })
        .sort((a, b) => sort === 'amount' ? a.amount - b.amount : a.createdAt - b.createdAt)
}

export const getTotalAmount = (expenses = []) => expenses.reduce((total, expense) => total + expense.amount, 0)