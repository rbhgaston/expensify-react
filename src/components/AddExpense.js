import React from 'react';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';
import {connect} from 'react-redux'

function AddExpense(props){
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm onSubmit={(expense) => {
                props.dispatch(addExpense(expense))
                props.history.push('/')
            }}/>
        </div>
    )
}

export default connect()(AddExpense)

