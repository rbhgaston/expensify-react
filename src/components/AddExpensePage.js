import React from 'react';
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm';
import {startAddExpense, addExpense} from '../actions/expenses';

export function AddExpensePage(props){
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.startAddExpense(expense)
                    props.history.push('/')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

// if you pass mapDispatchToProps, you lose access to dispatch in props
export default connect(undefined, mapDispatchToProps)(AddExpensePage)

