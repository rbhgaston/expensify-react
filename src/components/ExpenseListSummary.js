import React from 'react'
import {connect} from 'react-redux'
import {getFilteredExpenses, getTotalAmount} from '../selectors/expenses';
import numeral from 'numeral'

export const ExpenseListSummary = (props) => {
    const selectedExpenses = getFilteredExpenses(props.expenses, props.filters)

    const length = selectedExpenses.length
    const expenseWord = length === 1 ? 'expense' : 'expenses'
    const total = numeral(getTotalAmount(selectedExpenses) / 100).format(	'$0,0.00')

    return (
        <h2>Selecting {length} {expenseWord} with total: {total}</h2>
    )
}

const mapStateToProps = state => ({expenses: state.expenses, filters: state.filters})


export default connect(mapStateToProps)(ExpenseListSummary)