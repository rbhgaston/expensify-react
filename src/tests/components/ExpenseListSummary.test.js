import React from 'react'
import {ExpenseListSummary} from '../../components/ExpenseListSummary';
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses';
import {defaultFilters} from '../fixtures/filters';

let wrapper;
beforeEach(() => {
   wrapper = shallow(<ExpenseListSummary expenses={expenses} filters={defaultFilters}/>)
})

test('render ExpenseListSummary with multiple expenses', () => {
    expect(wrapper).toMatchSnapshot()
})

 test('render ExpenseListSummary with 1 expense', () => {
     wrapper.setProps({expenses: [expenses[0]]})
     expect(wrapper).toMatchSnapshot()
 })