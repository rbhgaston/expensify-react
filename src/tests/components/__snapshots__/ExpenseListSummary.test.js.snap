// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseListSummary with 1 expense 1`] = `
<h2>
  Selecting 
  1
   
  expense
   with total: 
  $1.00
</h2>
`;

exports[`render ExpenseListSummary with multiple expenses 1`] = `
<h2>
  Selecting 
  3
   
  expenses
   with total: 
  $6.00
</h2>
`;
